src/uct/cuda/cuda_ipc/cuda_ipc_iface.c

```diff
@@ -234,61 +230,60 @@ static ucs_status_t uct_cuda_ipc_iface_query(uct_iface_h tl_iface,
     return UCS_OK;
 }
 
-static ucs_status_t
-uct_cuda_ipc_iface_flush(uct_iface_h tl_iface, unsigned flags,
-                         uct_completion_t *comp)
+
+static ucs_status_t uct_cuda_ipc_iface_flush(uct_iface_h tl_iface, unsigned flags,
+                                              uct_completion_t *comp)
 {
     uct_cuda_ipc_iface_t *iface = ucs_derived_of(tl_iface, uct_cuda_ipc_iface_t);
+    uct_cuda_queue_desc_t *q_desc;
+    ucs_queue_iter_t iter;
 
     if (comp != NULL) {
         return UCS_ERR_UNSUPPORTED;
     }
 
-    if (ucs_queue_is_empty(&iface->outstanding_d2d_event_q)) {
-        UCT_TL_IFACE_STAT_FLUSH(ucs_derived_of(tl_iface, uct_base_iface_t));
-        return UCS_OK;
+    ucs_queue_for_each_safe(q_desc, iter, &iface->active_queue, queue) {
+        if (!ucs_queue_is_empty(&q_desc->event_queue)) {
+            UCT_TL_IFACE_STAT_FLUSH_WAIT(ucs_derived_of(tl_iface,
+                                                        uct_base_iface_t));
+            return UCS_INPROGRESS;
+        }
     }
 
-    UCT_TL_IFACE_STAT_FLUSH_WAIT(ucs_derived_of(tl_iface, uct_base_iface_t));
-    return UCS_INPROGRESS;
+    UCT_TL_IFACE_STAT_FLUSH(ucs_derived_of(tl_iface, uct_base_iface_t));
+    return UCS_OK;
+
 }
 
 static UCS_F_ALWAYS_INLINE unsigned
-uct_cuda_ipc_progress_event_q(uct_cuda_ipc_iface_t *iface,
-                              ucs_queue_head_t *event_q)
+uct_cuda_ipc_progress_event_queue(uct_cuda_ipc_iface_t *iface,
+                                   ucs_queue_head_t *queue_head,
+                                   unsigned max_events)
 {
     unsigned count = 0;
-    uct_cuda_ipc_event_desc_t *cuda_ipc_event;
-    ucs_queue_iter_t iter;
+    uct_cuda_ipc_event_desc_t *cuda_event;
     ucs_status_t status;
-    unsigned max_events = iface->config.max_poll;
 
-    ucs_queue_for_each_safe(cuda_ipc_event, iter, event_q, queue) {
-        status = UCT_CUDADRV_FUNC_LOG_ERR(cuEventQuery(cuda_ipc_event->event));
-        if (UCS_INPROGRESS == status) {
-            continue;
-        } else if (UCS_OK != status) {
-            return status;
+    ucs_queue_for_each_extract(cuda_event, queue_head, queue,
+                               cuEventQuery(cuda_event->event) ==
+                                       CUDA_SUCCESS) {
+        ucs_queue_remove(queue_head, &cuda_event->queue);
```

Comment:
why need to remove? (extract is used)

Reply:
We want the queue to hold only active events. When there are no events in the queue, we can ignore that event queue and minimize polls.